hooks/pre-commit

```diff
@@ -1,29 +1,62 @@
 #!/usr/bin/env bash
 
-echo "Uncrustifying staged files..."
+((${BASH_VERSION%%.*} >= 4)) || { echo >&2 "$0: Error: Please upgrade Bash."; exit 1; }
 
-RELPATH=../..
-staged=$(git diff --name-only --cached)
+set -euo pipefail
+
+function uncrustify_file()
+{
+    local -r file="$1"
+    cp "${file}" "${file}.bak"
+    if "tools/codestyle/uncrustify_run_file.sh" "${file}"; then
+        rm "${file}.bak"
+    else
+        status=$?
+        mv "${file}.bak" "${file}"
+        rm -f "${file}.uncrustify" # remove possible temporary file
+        return ${status}
+    fi
+}
+
+unset pids cpp_files error_statuses error_files # create empty arrays
+
+echo >&2 "$0: Uncrustifying staged files..."
+
+staged_files="$(git diff --cached --name-only)"
 
-cd ./tools/codestyle/
 while IFS= read -r file; do
-	ext="${file##*.}"
-	if [ $ext != "cpp" ] && [ $ext != "h" ]; then
-		continue
-	fi
-	cp "$RELPATH/$file" "$RELPATH/$file.bak" 
-	if $( ./uncrustify_run_file.sh "$RELPATH/$file" ); then
-		mv "$RELPATH/$file.uncrustify" "$RELPATH/$file"
-		rm "$RELPATH/$file.bak"
-		git add "$RELPATH/$file"
-	else
-		echo $?
-		mv "$RELPATH/$file.bak" "$RELPATH/$file"
-		rm "$RELPATH/$file.uncrustify"
-		echo "Uncrustify failed for $file"
-	fi
-done <<< "$staged"
-
-echo "Finished uncrustifying"
-
-exit 0
\ No newline at end of file
+    ext="${file##*.}"
+    if [[ "${ext}" != "cpp" ]] && [[ "${ext}" != "h" ]]; then
```

Comment:
Sure, I'll add these extensions.

Reply:
Don't we also have some `.c` files (in Third Party code)

Reply:
Yes, but third-party files should better not be uncrustified. 

@shoogle That brings me to another idea: it would be great if the pre-commit hook would only run on non-third-party files. For example, the file `uncrustify_run.sh` contains a list of the directories inside which the automatic code style check runs: 
```sh
SRC_DIRS=(
    bww2mxml
    fonttools
    miditools
    mtest
    src
)
```

Reply:
Now implemented for `m` and `mm` files. `SRC_DIRS` has been moved into a shared file `globals.source`.